File: HealthCareAtHome/shared/codes.js
/* codes.js — secret-code / OTP generation and verification.
 *
 * Two codes govern a booking's lifecycle (requirements 4, 7, 8, 9):
 *   - startCode    : issued when a caregiver accepts. Shared with BOTH client
 *                    and caregiver. On arrival the client asks the caregiver
 *                    for it and verifies to begin the service.
 *   - completeCode : issued when the caregiver requests completion. Shared with
 *                    the client, who enters it (with rating + comments) to
 *                    officially close the booking.
 *
 * Codes are numeric OTP-style strings. Verification is constant-time-ish and
 * whitespace/format tolerant.
 */

/** Generate a numeric code of `len` digits (default 6). */
export function generateCode(len = 6) {
  const max = 10 ** len;
  const n = (crypto.getRandomValues(new Uint32Array(1))[0] % max);
  return String(n).padStart(len, '0');
}

/** Normalise user input before comparing (strip spaces/dashes). */
export function normalize(input) {
  return String(input == null ? '' : input).replace(/[\s-]/g, '');
}

/** Length-safe equality check. */
export function verifyCode(expected, provided) {
  const a = normalize(expected);
  const b = normalize(provided);
  if (!a || a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) diff |= a.charCodeAt(i) ^ b.charCodeAt(i);
  return diff === 0;
}

File: HealthCareFlutter/mobile/packages/hc_core/lib/src/util/codes.dart
import 'dart:math';

/// Secret-code / OTP generation and verification (ports the PWA's codes.js).
///
/// Two codes govern a booking's lifecycle:
///   - startCode    : issued on caregiver accept, shared with both sides.
///   - completeCode : issued when the caregiver requests completion.
class Codes {
  static final _rand = Random.secure();

  /// Generate a numeric code of [len] digits (default 6).
  static String generate([int len = 6]) {
    final max = pow(10, len).toInt();
    final n = _rand.nextInt(max);
    return n.toString().padLeft(len, '0');
  }

  /// Normalise input before comparing (strip spaces/dashes).
  static String normalize(String? input) =>
      (input ?? '').replaceAll(RegExp(r'[\s-]'), '');

  /// Length-safe equality check.
  static bool verify(String? expected, String? provided) {
    final a = normalize(expected);
    final b = normalize(provided);
    if (a.isEmpty || a.length != b.length) return false;
    var diff = 0;
    for (var i = 0; i < a.length; i++) {
      diff |= a.codeUnitAt(i) ^ b.codeUnitAt(i);
    }
    return diff == 0;
  }
}
